lib/honeybadger/insights/base.ex

```diff
@@ -0,0 +1,175 @@
+      """
+      def dependencies_available? do
+        # We are using this strategy to deal with a compiler warning about the
+        # empty case never being reached.
+        deps = @required_dependencies
+
+        if Enum.empty?(deps) do
+          true
+        else
+          Enum.all?(deps, fn mod -> Code.ensure_loaded?(mod) end)
+        end
+      end
+
+      def event_filter(map, meta, name) do
+        if Application.get_env(:honeybadger, :event_filter) do
+          Application.get_env(:honeybadger, :event_filter).filter(map, meta, name)
+        else
+          map
+        end
+      end
+
+      def get_telemetry_events() do
+        get_insights_config(:telemetry_events, @telemetry_events)
+      end
+
+      def get_insights_config(key, default) do
+        insights_config = Application.get_env(:honeybadger, :insights_config, %{})
+        module_config = Map.get(insights_config, @config_namespace, %{})
+        Map.get(module_config, key, default)
+      end
+
+      @doc """
+      Attaches telemetry handlers if all required dependencies are available.
+      """
+      def attach do
+        if dependencies_available?() do
+          get_telemetry_events()
+          |> Enum.each(&attach_event/1)
+
+          :ok
+        else
+          Logger.debug(
+            "[Honeybadger] Missing Insights dependencies for #{inspect(@required_dependencies)}"
+          )
+
+          {:error, :missing_dependencies}
+        end
+      end
+
+      @doc """
+      Attaches a telemetry handler for a specific event.
+      The event can be either a string (dot-separated) or a list of atoms.
+      """
+      def attach_event(event) when is_binary(event) do
+        event_atoms = event |> String.split(".") |> Enum.map(&String.to_atom/1)
+        attach_event(event_atoms, event)
+      end
+
+      def attach_event(event) when is_list(event) do
+        event_name = Honeybadger.Utils.dotify(event)
+        attach_event(event, event_name)
+      end
+
+      defp attach_event(event_atoms, event_name) do
+        :telemetry.attach(
+          # Use the event name as the handler ID
+          event_name,
+          event_atoms,
+          &__MODULE__.handle_telemetry/4,
+          nil
+        )
+      end
+
+      defp process_measurements(measurements) do
+        measurements
+        |> Map.drop([
+          :monotonic_time,
+          # Absinthe
+          :end_time_mono
+        ])
+        |> Enum.reduce(%{}, fn {key, value}, acc ->
+          case key do
+            key
+            when key in [
+                   :duration,
+                   :total_time,
+                   :decode_time,
+                   :query_time,
+                   :queue_time,
+                   :idle_time
+                 ] ->
+              Map.put(acc, key, System.convert_time_unit(value, :native, :millisecond))
+
+            _ ->
+              Map.put(acc, key, value)
+          end
+        end)
```

Comment:
Can be done with a single reduce to avoid iterating through the measurements twice (but I extracted `@time_keys` into a module attribute)

```suggestion
        Enum.reduce(measurements, %{}, fn
          {key, _vl}, acc when key in ~w(monotonic_time end_time_mono)a ->
            acc

          {key, val}, acc when key in @time_keys ->
            Map.put(acc, key, System.convert_time_unit(val, :native, :millisecond))

          {key, val}, acc ->
            Map.put(acc, key, val)
        end)
```